include/ofi_util.h

```diff
@@ -1328,15 +1328,34 @@ static inline void ofi_cq_err_memcpy(uint32_t api_version,
 				     const struct fi_cq_err_entry *prov_buf)
 {
 	size_t size;
+	char *err_buf_save;
+	size_t err_data_size;
 
-	if (FI_VERSION_GE(api_version, FI_VERSION(1, 20)))
-		size = sizeof(struct fi_cq_err_entry);
-	else if (FI_VERSION_GE(api_version, FI_VERSION(1, 5)))
-		size = sizeof(struct fi_cq_err_entry_1_1);
-	else
-		size = sizeof(struct fi_cq_err_entry_1_0);
+	if (FI_VERSION_LT(api_version, FI_VERSION(1, 5))) {
```

Comment:
@j-xiong : Is this commit causing the Intel CI problems? If so, I am OK with removing this commit and keeping this logic in util CQ code.

Reply:
I haven't got to the root cause yet, but the `shm` provider test has been failing consistently with this PR.

Reply:
There are four failed tests with the shm provider: `fi_av_xfer`, `fi_rdm_tagged_peek`, `fi_multi_recv`, and `fi_ubertest`. The output from these tests looks normal, but the result is set to fail.  Since the shm provider uses util_cq, these is very likely related to this PR. Example in the log:

```
11:28:05  - name:   fi_av_xfer -e rdm -p "shm"
11:28:05    timestamp: 20231026-182306+0000
11:28:05    result: Fail
11:28:05    time:   4
11:28:05    server_cmd:  fi_av_xfer -e rdm -p "shm"   -s cb7
11:28:05    server_stdout: |
11:28:05      AV address removal: PASS
11:28:05      AV re-insertion address: PASS
11:28:05    client_cmd:  fi_av_xfer -e rdm -p "shm"   -s cb7 cb7
11:28:05    client_stdout: |
```

```
11:28:05  - name:   fi_rdm_tagged_peek -p "shm"
11:28:05    timestamp: 20231026-182316+0000
11:28:05    result: Fail
11:28:05    time:   3
11:28:05    server_cmd:  fi_rdm_tagged_peek -p "shm"   -s cb7
11:28:05    server_stdout: |
11:28:05    client_cmd:  fi_rdm_tagged_peek -p "shm"   -s cb7 cb7
11:28:05    client_stdout: |
11:28:05      Peek for a bad msg
11:28:05      Peek w/ claim for a bad msg
11:28:05      Peek msg 1
11:28:05      Receive msg 1
11:28:05      Peek w/ claim msg 2
11:28:05      Receive claimed msg 2
11:28:05      Peek & discard msg 3
11:28:05      Checking to see if msg 3 was discarded
11:28:05      Peek w/ claim msg 4
11:28:05      Claim and discard msg 4
11:28:05      Receive msg 5
11:28:05      Receive msg 6
11:28:05      Receive msg 10
11:28:05      Receive msg 9
11:28:05      Receive msg 8
11:28:05      Receive msg 7
```

Reply:
I've tracked it to a segfault writing entries on close for the canceled receives. I can look into it.
```
#0  0x00007ffff6c8e144 in __memmove_avx_unaligned_erms ()
   from /lib64/libc.so.6
#1  0x00007ffff7a45283 in mem_dup (src=0x7ffff7a53b3f <slist_remove_head+55>, 
    size=6615392) at ./include/ofi_mem.h:84
#2  0x00007ffff7a45d9e in util_cq_insert_error (cq=0x630f70, 
    err_entry=0x7fffffffd990) at prov/util/src/util_cq.c:99
#3  0x00007ffff7a471b1 in util_peer_cq_writeerr (cq=0x622820, 
    err_entry=0x7fffffffd990) at prov/util/src/util_cq.c:589
#4  0x00007ffff7a54344 in ofi_peer_cq_write_error (cq=0x630f70, 
    err_entry=0x7fffffffd990) at ./include/ofi_util.h:697
#5  0x00007ffff7a566c1 in util_cancel_entry (srx=0x64f0b0, flags=2056, 
    rx_entry=0x7ffff7eb0140) at prov/util/src/util_srx.c:891
#6  0x00007ffff7a56893 in util_srx_close (fid=0x64f0b0)
    at prov/util/src/util_srx.c:942
#7  0x00007ffff7b05662 in smr_ep_close (fid=0x64eb20)
    at prov/shm/src/smr_ep.c:835
#8  0x0000000000402446 in fi_close (fid=0x64eb20)
    at /home/aingerso/install/include/rdma/fabric.h:633
#9  0x0000000000408911 in ft_close_fids () at common/shared.c:1716
#10 0x00000000004096ba in ft_free_res () at common/shared.c:1786
#11 0x0000000000401e2f in av_removal_test () at functional/av_xfer.c:121
#12 0x00000000004023dc in main (argc=6, argv=0x7fffffffdca8)
    at functional/av_xfer.c:243
   ```

Reply:
the `size` parameter in the `mem_dup` call doesn't look right.

Reply:
Ah you're totally right. The srx code is not properly initializing the err data size:

```
static int util_cancel_entry(struct util_srx_ctx *srx, uint64_t flags,
			     struct util_rx_entry *rx_entry)
{
	struct fi_cq_err_entry err_entry;
	int ret;

	err_entry.op_context = rx_entry->peer_entry.context;
	err_entry.flags = flags;
	err_entry.tag = rx_entry->peer_entry.tag;
	err_entry.err = FI_ECANCELED;
	err_entry.prov_errno = -FI_ECANCELED;

	ret = ofi_peer_cq_write_error(srx->cq, &err_entry);
	ofi_buf_free(rx_entry);
	return ret ? ret : 1;
}
```
Probably should add an initialization in the err_entry to {0}

Reply:
I think here is the culprit (prov/util/src/util_srx.c): some fields (including `err_data` and `err_data_size`) of `err_entry` is uninitialized. 

```
static int util_cancel_entry(struct util_srx_ctx *srx, uint64_t flags,
                             struct util_rx_entry *rx_entry)
{
        struct fi_cq_err_entry err_entry;
        int ret;

        err_entry.op_context = rx_entry->peer_entry.context;
        err_entry.flags = flags;
        err_entry.tag = rx_entry->peer_entry.tag;
        err_entry.err = FI_ECANCELED;
        err_entry.prov_errno = -FI_ECANCELED;

        ret = ofi_peer_cq_write_error(srx->cq, &err_entry);
        ofi_buf_free(rx_entry);
        return ret ? ret : 1;
}
```

@iziemba Could you add a commit to fix this issue? Just need to add `= { 0 };` to the declaration of `err_entry`.

Reply:
@aingerson Sorry for the duplication, I didn't refresh window before posting follow-up comments.

Reply:
Yes. I can fix this.